packages/flutter/lib/src/widgets/editable.dart

```diff
@@ -430,3 +463,153 @@ class _EditableLineWidget extends LeafRenderObjectWidget {
+
+    TextSelection newSelection;
+    if (side == _TextSelectionHandleSide.left) {
+      newSelection = new TextSelection(
+        baseOffset: position.offset,
+        extentOffset: selection.extentOffset
+      );
+    } else {
+      newSelection = new TextSelection(
+        baseOffset: selection.baseOffset,
+        extentOffset: position.offset
+      );
+    }
+
+    if (newSelection.baseOffset >= newSelection.extentOffset)
+      return;
+
+    setState(() {
+      selection = newSelection;
+      endpoints = config.renderObject.getEndpointsForSelection(selection);
+      config.onSelectionHandleChanged(selection);
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return new Stack(
+      children: <Widget>[
+        new GestureDetector(
+          onHorizontalDragStart: (Point p) { _dragStart(p, _TextSelectionHandleSide.left); },
+          onHorizontalDragUpdate: (double d) { _dragUpdate(d, _TextSelectionHandleSide.left); },
+          child: new CustomPaint(
+            foregroundPainter: new _TextSelectionHandlePainter(
+              side: _TextSelectionHandleSide.left,
+              position: endpoints[0]
+            )
+          )
+        ),
+        new GestureDetector(
+          onHorizontalDragStart: (Point p) { _dragStart(p, _TextSelectionHandleSide.right); },
+          onHorizontalDragUpdate: (double d) { _dragUpdate(d, _TextSelectionHandleSide.right); },
+          child: new CustomPaint(
+            foregroundPainter: new _TextSelectionHandlePainter(
+              side: _TextSelectionHandleSide.right,
+              position: endpoints[1]
+            )
+          )
+        )
+      ]
+    );
+  }
+}
+
+/// Draws a single text selection handle at a fixed screen position.
+class _TextSelectionHandlePainter extends CustomPainter {
+  static const double _kSelectionHandleRadius = 10.0; // pixels
+
+  _TextSelectionHandlePainter({this.side, this.position});
+
+  final Point position;
+  final _TextSelectionHandleSide side;
+
+  // Returns a rect of the given size at the corner of the selection range.
+  Rect _getCornerRect(double size) {
+    if (side == _TextSelectionHandleSide.left)
+      return new Rect.fromLTWH(position.x - size, position.y, size, size);
+    else
+      return new Rect.fromLTWH(position.x, position.y, size, size);
+  }
+
+  @override
+  void paint(Canvas canvas, Size size) {
+    // TODO(mpcomplete): pick a better color. Use an accent color based on the
+    // selection color?
+    Paint paint = new Paint()..color = new Color(0xff0000ff);
+
+    // Each handle is a circle, with a rectangle in the top quadrant of that
+    // circle in the direction it's pointing. [rect] here is the size of the
+    // corner rect, e.g. half the diameter of the circle.
+    Rect rect = _getCornerRect(_kSelectionHandleRadius);
+    canvas.drawCircle(
+      (side == _TextSelectionHandleSide.left) ? rect.bottomLeft : rect.bottomRight,
+      _kSelectionHandleRadius, paint);
```

Comment:
These are material-design selection handles.  Other design languages will want to use different styles of selection handles.  We'll need to factor the code such that `Input` (which is in the material layer) decides how the selection handles should appear.